src/catalog/src/durable/impls/persist.rs

```diff
@@ -238,48 +267,62 @@ impl PersistHandle {
+    /// `upper`.
+    fn as_of(&self, upper: Timestamp) -> Timestamp {
+        as_of(&self.read_handle, upper)
+    }
+
     /// Reports if the catalog state has been initialized, and the current upper.
     async fn is_initialized_inner(&mut self) -> (bool, Timestamp) {
         let upper = self.current_upper().await;
-        (upper > 0.into(), upper)
+        (upper > *SENTINEL_UNINITIALIZED, upper)
+    }
+
+    /// Generates a [`Vec<StateUpdate>`] that contain all updates to the catalog
+    /// state.
+    ///
+    /// The output is consolidated and sorted by timestamp in ascending order and the current upper.
+    #[tracing::instrument(level = "debug", skip(self))]
+    async fn current_snapshot(&mut self) -> (Vec<StateUpdate>, Timestamp) {
+        let (is_initialized, current_upper) = self.is_initialized_inner().await;
+        if is_initialized {
+            let as_of = self.as_of(current_upper);
+            let snapshot = self.snapshot(as_of).await.collect();
+            (snapshot, current_upper)
+        } else {
+            (Vec::new(), current_upper)
+        }
     }
 
     /// Generates an iterator of [`StateUpdate`] that contain all updates to the catalog
-    /// state up to, but not including, `upper`.
+    /// state up to, and including, `as_of`.
     ///
     /// The output is consolidated and sorted by timestamp in ascending order.
     #[tracing::instrument(level = "debug", skip(self))]
     async fn snapshot(
         &mut self,
-        upper: Timestamp,
+        as_of: Timestamp,
     ) -> impl Iterator<Item = StateUpdate> + DoubleEndedIterator {
-        snapshot(&mut self.read_handle, upper).await
+        snapshot(&mut self.read_handle, as_of).await
     }
 
     /// Generates an iterator of [`StateUpdate`] that contain all updates to the catalog
-    /// state up to, but not including, `upper`.
+    /// state up to, and including, `as_of`.
     async fn snapshot_unconsolidated(
         &mut self,
-        upper: Timestamp,
+        as_of: Timestamp,
     ) -> impl Iterator<Item = StateUpdate> + DoubleEndedIterator {
-        let snapshot = if upper > Timestamp::minimum() {
-            let since = self.read_handle.since().clone();
-            let mut as_of = upper.saturating_sub(1);
-            as_of.advance_by(since.borrow());
-            let mut snapshot = Vec::new();
-            let mut stream = Box::pin(
-                self.read_handle
-                    .snapshot_and_stream(Antichain::from_elem(as_of))
-                    .await
-                    .expect("we have advanced the restart_as_of by the since"),
-            );
-            while let Some(update) = stream.next().await {
-                snapshot.push(update)
-            }
-            snapshot
-        } else {
-            Vec::new()
-        };
+        let mut snapshot = Vec::new();
+        let mut stream = Box::pin(
+            self.read_handle
+                .snapshot_and_stream(Antichain::from_elem(as_of))
```

Comment:
I realize this is not new code, but I forget why do we use `snapshot_and_stream` here instead of `snapshot_and_fetch`? Maybe something about not needing it to be consolidated? We should leave a comment here on why, plz